apisix/discovery/k8s.lua

```diff
@@ -0,0 +1,601 @@
+            resource.watch_state = "listing"
+            resource:pre_list_callback()
+            ok, reason, message = list_resource(httpc, resource, nil)
+            if not ok then
+                resource.watch_state = "list failed"
+                core.log.error("list failed, resource: ", resource.plural,
+                        ", reason: ", reason, ", message : ", message)
+                retry_interval = 100
+                break
+            end
+            resource.watch_state = "list finished"
+            resource:post_list_callback()
+
+            core.log.info("begin to watch ", resource.plural)
+            resource.watch_state = "watching"
+            ok, reason, message = watch_resource(httpc, resource)
+            if not ok then
+                resource.watch_state = "watch failed"
+                core.log.error("watch failed, resource: ", resource.plural,
+                        ", reason: ", reason, ", message : ", message)
+                retry_interval = 0
+                break
+            end
+            resource.watch_state = "watch finished"
+            retry_interval = 0
+        until true
+
+        -- every 3 hours,we should quit and use another timer
+        local now_time = ngx.time()
+        if now_time - begin_time >= 10800 then
+            break
+        end
+        if retry_interval ~= 0 then
+            ngx.sleep(retry_interval)
+        end
+    end
+    local timer_runner = function()
+        fetch_resource(resource)
+    end
+    ngx.timer.at(0, timer_runner)
+end
+
+local function create_endpoint_lrucache(endpoint_key, endpoint_port)
+    local endpoint_content, _, _ = endpoints_shared:get_stale(endpoint_key)
+    if not endpoint_content then
+        core.log.emerg("get empty endpoint content from discovery DICT,this should not happen ",
+                endpoint_key)
+        return nil
+    end
+
+    local endpoint, _ = core.json.decode(endpoint_content)
+    if not endpoint then
+        core.log.emerg("decode endpoint content failed, this should not happen, content : ",
+                endpoint_content)
+    end
+
+    return endpoint[endpoint_port]
+end
+
+local host_patterns = {
+    { pattern = [[^\${[_A-Za-z]([_A-Za-z0-9]*[_A-Za-z])*}$]] },
+    { pattern = [[^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$]] },
+}
+
+local port_patterns = {
+    { pattern = [[^\${[_A-Za-z]([_A-Za-z0-9]*[_A-Za-z])*}$]] },
+    { pattern = [[^(([1-9]\d{0,3}|[1-5]\d{4}|6[0-4]\d{3}|65[0-4]\d{2}|655[0-2]\d|6553[0-5]))$]] },
+}
+
+local schema = {
+    type = "object",
+    properties = {
+        service = {
+            type = "object",
+            properties = {
+                schema = {
+                    type = "string",
+                    enum = { "http", "https" },
+                    default = "https",
+                },
+                host = {
+                    type = "string",
+                    default = "${KUBERNETES_SERVICE_HOST}",
```

Comment:
Are you trying to get the environment variables directly in the schema here?

Reply:
my means 
1. set default value is “${ KUBERNETES_SERVER_HOST}” in schema

> 2 wiil call os.getenv("KUBERNETES_SERVER_HOST") in code

Reply:
why not just use `"KUBERNETES_SERVER_HOST"` as default value in schema, and use `os.getenv("KUBERNETES_SERVER_HOST")` in code

Reply:
I understand what you mean, but it is inappropriate.

User can set the host to any suitable value, maybe an IP, maybe a domain name, or an environment variable
If we don't use "${}" as the identifier, we won't know the exact meaning of the user.